File: 3_Shared_Memory_Systems/WinAPI/vector.cpp
#include "stdafx.h"

#include <iostream>
#include <cstdlib>

using namespace std;

Vector::Vector(int size, int number)
{
	_size = size;
	_data = new int[size];

	for (int i = 0; i < size; i++)
			_data[i] = number;
}

Vector::Vector(Vector* vector)
{
	_size = vector->getSize();
	_data = new int[_size];

	for (int i = 0; i < _size; i++)
		_data[i] = vector->getElement(i);
}

Vector::~Vector()
{
	delete[] _data;
}

void Vector::sort(int startIndex, int finishIndex)
{
	int tmp;
	for (int i = startIndex; i <= finishIndex; i++)
		for (int j = 1; j < _size - i; j++)
			if (_data[j - 1] > _data[j])
			{
				tmp = _data[j - 1];
				_data[j - 1] = _data[j];
				_data[j] = tmp;
			}
}

void Vector::Swap(int indexI, int indexJ)
{
	int buf = _data[indexI];
	_data[indexI] = _data[indexJ];
	_data[indexJ] = buf;
}

void Vector::mergeSort(int startIndex, int finishIndex)
{
	int h = (finishIndex - startIndex) / 2;
	for (int i = startIndex; i < (finishIndex + startIndex) / 2; i++)
		if (_data[i] > _data[i + h])
		{
			Swap(i, i + h);
		}
}

Vector* Vector::MulVectorNumber(int number)
{
	Vector* result = new Vector(_size, 0);
	for (int i = 0; i < _size; i++)
	{
		result->setElement(i, _data[i]*number);
	}
	return result;
}

void Vector::AddVectors(Vector* vector1, Vector* vector2, Vector* vector3, int up, int down)
{
	for (int i = 0; i < _size; i++)
		_data[i] = vector1->getElement(i) + vector2->getElement(i) + vector3->getElement(i);
}


void Vector::setElement(int index, int value)
{
	_data[index] = value;
}

void Vector::print()
{
	if (_size <= 10)
		for (int i = 0; i < _size; i++)
			printf("%6d ", _data[i]);
	else printf("Output is to big!");
	printf("\n");
}

int Vector::getElement(int index)
{
	return _data[index];
}

int Vector::getSize()
{
	return _size;
}

File: 3_Shared_Memory_Systems/Java/ScalableSystemWithSharedMemory/src/com/kirintor/classes/Vector.java
package com.kirintor.classes;

import java.util.ArrayList;
import java.util.Random;

public class Vector {

	private int n;
	private int[] data;
	
	public Vector(){		
	}
	
	public Vector(int n, int number){
		data = new int[n];
        this.n = n;
        for (int i = 0; i < n; i++)
        {
        	data[i] = number;
        }
	}
	
	public Vector(int n, boolean isRandom) {
        this.n = n;
        this.data = new int[n];
        if (isRandom) {
            Random rnd = new Random();
            for (int i = 0; i < n; i++) {
                this.data[i] = rnd.nextInt(100);
            }
        }
    }
	
	public Vector(int n) {
        this.n = n;
        this.data = new int[n];
    }
	
	public Vector(Vector vector){
		this.n = vector.getN();
		data = new int[n];
		for(int i =0; i<n; i++)
		{
			data[i] = vector.getElement(i);
		}
	}
	
	public Vector(Vector vector, int startIndex, int finishIndex){
		this.n = finishIndex-startIndex;
		data = new int[n];
		for(int i =0; i<n; i++)
		{
			data[i] = vector.getElement(startIndex);
			startIndex++;
		}
	}
	
	public Vector sort(){
		int tmp;
		for (int i = 0; i < n; i++)
			for (int j = 1; j < n - i; j++)
				if (data[j - 1] > data[j])
				{
					tmp = data[j - 1];
					data[j - 1] = data[j];
					data[j] = tmp;
				}
		return this;
	}
	
	public void sort(int startIndex, int finishIndex){
		int tmp;
		for (int i = startIndex; i < finishIndex; i++)
			for (int j = 1; j < n - i; j++)
				if (data[j - 1] > data[j])
				{
					tmp = data[j - 1];
					data[j - 1] = data[j];
					data[j] = tmp;
				}
	}

	@SuppressWarnings("unused")
	private void swap(int indexI, int indexJ){
		int buf = data[indexI];
		data[indexI] = data[indexJ];
		data[indexJ] = buf;
	}

	public Vector mergeSort(Vector vector1, Vector vector2){
		Vector result = new Vector(vector1.getN()+vector2.getN(), 0);
		ArrayList<Integer> array1 = new ArrayList<Integer>(vector1.getN());
		ArrayList<Integer> array2 = new ArrayList<Integer>(vector2.getN());
		ArrayList<Integer> resultList = new ArrayList<Integer>(result.getN());
		
		for(int i = 0; i<vector1.getN(); i++){
			array1.add(Integer.valueOf(vector1.data[i]));
		}
		for(int j = 0; j<vector2.getN(); j++){
			array2.add(Integer.valueOf(vector2.data[j]));
		}
		
		while(((!array1.isEmpty())
			 &&(!array2.isEmpty()))){
			if (array1.get(0) <= array2.get(0)){
				resultList.add(array1.get(0));
				array1.remove(0);
			}else{
				resultList.add(array2.get(0));
				array2.remove(0);
			}
		}
		if(array1.isEmpty()){
			while(!array2.isEmpty()){
				resultList.add(array2.get(0));
				array2.remove(0);
			}
		}else{
			while(!array1.isEmpty()){
				resultList.add(array1.get(0));
				array1.remove(0);
			}
		}
		
		for(int i = 0; i<result.getN(); i++){
			result.setElement(resultList.get(i), i);
		}
		
		return result;
	}
	
	public void print(){
        if (n <= 20){
        	System.out.print("[");
            for (int i = 0; i < n-1; i++){
            	System.out.print(data[i] + ", ");
            }
            System.out.println(data[n-1]+"]");
        }
        else{
        	System.out.println("Output is too big!");
        }
    }
	
	public String toString(){
		String result = "[";
		for (int i = 0; i < n-1; i++)
        {
        	result += data[i] + ", ";
        }
		result += data[n-1]+"]";
		return result;		
	}

	public int getN() {
		return n;
	}

	public void setN(int n) {
		this.n = n;
	}
	
	public int[] getData()
	{
		return this.data;
	}

	public int getElement(int index) {
		return data[index];
	}

	public void setElement(int element, int index) {
		data[index] = element;
	}
	
}
